### PR TITLE
read in all possible constituents from file, not just build-time ones

### PR DESCRIPTION
**Pull request by peverwhee**
Tag name (required for release branches):
Originator(s): peverwhee

Summary (include the keyword ['closes', 'fixes', 'resolves'] and issue number):
- Updates to enable reading in data for runtime constituents

Describe any changes made to build system:
M    cime_config/cam_autogen.py
M    cime_config/cam_build_cache.py
M    cime_config/cam_config.py
- keep track of registry constituents

Describe any changes made to the namelist: None

List any changes to the defaults for the input datasets (e.g. boundary datasets): None

List all files eliminated and why: None

List all files added and what they do: None

List all existing files that have been modified, and describe the changes: 
(Helpful git command: `git diff --name-status development...<your_branch_name>`)
M    src/data/generate_registry_data.py
- create list of registry constituents to pass to write_init_files
M    src/data/write_init_files.py
- handle constituents separately from other host variables; read/check ALL constituent variables, not just "required" ones
M    test/unit/test_*.py
- update calling lists

If there are new failures (compare to the `test/existing-test-failures.txt` file),
have them OK'd by the gatekeeper, note them here, and add them to the file.
If there are baseline differences, include the test and the reason for the
diff. What is the nature of the change? Roundoff?

derecho/intel/aux_sima:

derecho/gnu/aux_sima:

If this changes climate describe any run(s) done to evaluate the new
climate in enough detail that it(they) could be reproduced:

CAM-SIMA date used for the baseline comparison tests if different than latest:

### CODE REVIEW

**Review by nusbaume — COMMENTED** [review]
Looks good, but did have a few questions!

**Review by cacraigucar — APPROVED** [review]
Since this is mainly python, I have restricted my review to using the actual implementation.   I've incorporated this code into my ZM testing, and it fixed the problem that I encountered which led to this PR

File: src/data/write_init_files.py [hunk 1]
```diff
@@ -550,6 +573,17 @@ def write_ic_arrays(outfile, ic_name_dict, ic_max_len,
         #Write line to file:
         outfile.write(log_arr_str, 2)
     # end for
+    for var_num, var_name in enumerate(registry_constituents):
```
[the remark this answers is not in the document]
Comment (peverwhee) [reply src/data/write_init_files.py:577]: good call! added

File: src/data/write_init_files.py [hunk 2]
```diff
@@ -579,6 +612,16 @@ def write_ic_arrays(outfile, ic_name_dict, ic_max_len,
         #Write line to file:
         outfile.write(log_arr_str, 2)
     # end for
+    for var_num, varname in enumerate(registry_constituents):
```
[the remark this answers is not in the document]
Comment (peverwhee) [reply src/data/write_init_files.py:617]: reset!

File: src/data/write_init_files.py [hunk 3]
```diff
@@ -1072,6 +1078,60 @@ def write_phys_read_subroutine(outfile, host_dict, host_vars, host_imports,
...
+    outfile.write("call const_props(constituent_idx)%default_value(constituent_default_value, constituent_errflg, constituent_errmsg)", 5)
+    outfile.write("if (constituent_errflg /= 0) then", 5)
+    outfile.write("call endrun(constituent_errmsg, file=__FILE__, line=__LINE__)", 6)
+    outfile.write("end if", 5)
+    outfile.write("field_data_ptr(:,:,constituent_idx) = constituent_default_value", 5)
+    outfile.write("if (masterproc) then", 5)
+    outfile.write("write(iulog,*) 'Constituent ', trim(std_name), ' initialized to default value: ', constituent_default_value", 6)
+    outfile.write("end if", 5)
+    outfile.write("else", 4)
+    outfile.write("field_data_ptr(:,:,constituent_idx) = 0._kind_phys", 5)
```
[the remark this answers is not in the document]
Comment (peverwhee) [reply src/data/write_init_files.py]: yep - changed!

File: src/data/write_init_files.py [hunk 4]
```diff
@@ -1072,6 +1078,60 @@ def write_phys_read_subroutine(outfile, host_dict, host_vars, host_imports,
...
+    outfile.comment("Read in constituent variables if not using init variables", 2)
+    outfile.write("field_data_ptr => cam_constituents_array()", 2)
+    outfile.write("const_props => cam_model_const_properties()", 2)
+    outfile.blank_line()
+    outfile.comment("Iterate over all registered constituents", 2)
+    outfile.write("do constituent_idx = 1, size(const_props)", 2)
+    outfile.write("var_found = .false.", 3)
+    outfile.comment("Check if constituent standard name in registered SIMA standard names list:", 3)
+    outfile.write("call const_props(constituent_idx)%standard_name(std_name)", 3)
+    outfile.write("if(any(phys_var_stdnames == trim(std_name))) then", 3)
```
[the remark this answers is not in the document]
Comment (peverwhee) [reply src/data/write_init_files.py:1095]: I don't think so but I wasn't able to find any fortran documentation that supports my feeling! I did ask chatgpt and it said:

> if phys_var_stdnames contains strings with trailing spaces, they will still match trim(std_name) as long as the trimmed portion of std_name matches the untrimmed part of an element in phys_var_stdnames.

**Comment by peverwhee** [comment]
This PR is waiting on a bugfix in the framework and also https://github.com/ESCOMP/atmospheric_physics/pull/149

File: src/data/write_init_files.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cacraigucar) [reply src/data/write_init_files.py:1095]: I agree with @peverwhee and chatgpt, that this should be fine.

**Review by nusbaume — APPROVED** [review]

**Pull request merged**